WIP Use framework-specific dependencies for .Net Framework versions

iskiselev opened the pull request:
## Why
Currently OTEL registers in GAC assemblies, that was never intendend to be put in GAC on latest framework - and not registring netstandard on 4.6.2. It result in issues discussed at https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/4186#issuecomment-2979050475

## What
With this change, we pack different set of assemblies for supported .NET Frameworks:
- 4.6.2
- 4.7
- 4.7.1
- 4.7.2 and up

Only assemblies for current framework will be registered in GAC at install time or loaded from installation dir.
To minimize size set of assemblies deduplicated - with assemblies that used in all .NET Frameworks still stay at `netfx` folders. Assemblies used only in some versions stays in oldest TFM folder. For other TFM using same assembly .link file will be created that points to TFM with assembly.
To simplify buget packages tracking `CentralPackageTransitivePinningEnabled` was enabled - with it transient auto-generated packages were removed from `OpenTelemetry.AutoInstrumentation.csproj`.
.NET Framework version resolved though registry check in installation, profiler and managed code. It associated with some risk (proper rights). If version will not be resolved, it is fallback to 4.6.2, which should have same behaviour as before this change. It may be better nowdays to fallback to 4.7.2 - as most people probably have it.

## Tests
- validated OTEL work with loading assemblies from install folder on .NET 4.7.2. Manually validated that .link files are loaded correctly
- TODO: validate that GAC installation works properly
- TODO: validate that nuget installation was not affected

## Checklist

- [ ] `CHANGELOG.md` is updated.
- [ ] Documentation is updated.
- [ ] New features are covered by tests.

RassK commented on build/Build.Steps.cs:
```diff
@@ -42,9 +44,17 @@ partial class Build
     private static readonly IEnumerable<TargetFramework> TargetFrameworks = new[]
+    {
+       TargetFramework.NET8_0,
+       TargetFramework.NET462,
```
`TargetFramework.NET462` should this remain here? Becomes quite confusing between `TargetFrameworks` vs `TargetFrameworksNetFx `

Kielek replied:
`TargetFrameworksForNetFxRedirections`?

nrcventura replied:
`TargetFrameworksForNetFxPacking`?

martincostello commented on src/OpenTelemetry.AutoInstrumentation.Native/cor_profiler.cpp:
```diff
@@ -3643,4 +3650,98 @@ HRESULT STDMETHODCALLTYPE CorProfiler::EventPipeEventDelivered(EVENTPIPE_PROVIDE
+            else
+            {
+                Logger::Warn(
+                    "DetectFrameworkVersionTableForRedirectsMap: Old .Net Framework detected, use 462 as fallback");
```
```suggestion
                    "DetectFrameworkVersionTableForRedirectsMap: Old .NET Framework detected, use 462 as fallback");
```

iskiselev commented on build/TargetFramework.cs:
```diff
@@ -9,12 +9,16 @@ public class TargetFramework : Enumeration
     public static readonly TargetFramework NET462 = new() { Value = "net462" };
+    public static readonly TargetFramework NET47 = new() { Value = "net47" };
+    public static readonly TargetFramework NET471 = new() { Value = "net471" };
+    public static readonly TargetFramework NET472 = new() { Value = "net472" };
```
Currently we don't need net48 and net481, as all of our dependencies does not have difference between 4.7.2 and higher. I've added new `NetFrameworkDistroTests.ReferencedPackagesNoUnsupportedTfm` test that validates it. If any dependencies will do customizations for 4.8, 4.8.1 or potential newer .Net Framework - it will fail. There is comment in tests what should be done in that case.

nrcventura commented:
> More automation tests? Suggest what to cover.

One idea for a set of integration tests is having a test app that runs in windows containers with only a specific .net framework version available to make it easier to validate behavior for each specific .net framework version. I don't know what type of test app we need to have to ensure that the correct dependencies are in place. I also don't know if these tests will have enough value given how long it can take to download and run these containers.

Kielek commented on CHANGELOG.md:
```diff
@@ -39,6 +39,19 @@ This component adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.h
 - .NET Framework only, following packages updated
   - `OpenTelemetry.Instrumentation.AspNet` from `1.12.0-beta.1` to `1.13.0-beta.2`.
 
+#### Dependencies on .Net Framework
```
```suggestion
#### Dependencies on .NET Framework
```

Kielek commented on test/IntegrationTests/BuildTests.DistributionStructure_windows.verified.txt:
Could you please share the new/old package sizes on Windows?

iskiselev replied:
Od size 7,415,386 bytes (from #4560 windows build)
New size 7,838,414 bytes (from windows build of this PR after latest master merged)

So, windows zip size is increasing by 6% or 0.4 MB.

iskiselev commented on src/OpenTelemetry.AutoInstrumentation.Native/cor_profiler.cpp:
```diff
@@ -3628,4 +3635,108 @@ HRESULT STDMETHODCALLTYPE CorProfiler::EventPipeEventDelivered(EVENTPIPE_PROVIDE
+                // Map release numbers to framework versions
+                // Based on Microsoft documentation:
+                // https://docs.microsoft.com/en-us/dotnet/framework/migration-guide/how-to-determine-which-versions-are-installed
+                if (releaseValue >= 461308)
```
With new approach to detect 4.7.2 we will not use this branch on 4.7.2 and up at all. There is comment about it at lines 3644-3646.
I can either:
- extend that comment, specifying that newer framework version detection should be added to registry check
- add other frameworks here as commented code with words that it should not be used due to `ICorProfilerInfo8` check
- add detection with pinnig to 4.7.1 or 4.7.2

iskiselev replied:
Updated comment in front of if/else block that does ICorProfilerInfo8 check for 4.7.2 and up.

nrcventura commented on src/OpenTelemetry.AutoInstrumentation.Loader/AssemblyResolver.NetFramework.cs:
```diff
@@ -50,33 +41,91 @@ private static string ResolveManagedProfilerDirectory()
     }
+
+    [DllImport("OpenTelemetry.AutoInstrumentation.Native.dll")]
+    private static extern int GetDetectedNetFrameworkVersion();
```
Should we add a comment here that indicates that it will only detect versions up to 4.7.2?

iskiselev commented on test/IntegrationTests/BuildTests.DistributionStructure_windows.verified.txt:
```diff
@@ -83,110 +82,220 @@
+  \netfx\net471\System.Security.SecureString.dll,
+  \netfx\net471\System.Threading.Overlapped.dll,
+  \netfx\net471\System.Xml.XPath.XDocument.dll,
+  \netfx\net472\_._,
```
It is generated at https://github.com/iskiselev/opentelemetry-dotnet-instrumentation/blob/8a75f87ed123a481f585aaa6c42caf6cab23f68d/build/Build.Steps.cs#L445

Kielek commented on script-templates/OpenTelemetry.DotNet.Auto.psm1.template:
```diff
@@ -308,8 +307,36 @@ function Register-AssembliesInGAC() {
+        { $_ -ge 533320 } { $fxversion = 'net472'; break } # net481 or later, pin to 4.7.2
+        { $_ -ge 528040 } { $fxversion = 'net472'; break } # net48, pin to 4.7.2
+        { $_ -ge 461808 } { $fxversion = 'net472'; break }
+        { $_ -ge 461808 } { $fxversion = 'net472'; break }
```
```suggestion
        { $_ -ge 461808 } { $fxversion = 'net472'; break }
```

nrcventura commented:
I think the last remaining question on this PR is the following.

> > Decide if we need any replacement for deleted TransientDependenciesTests.DefinedTransientDeps_Are_MatchingGeneratedDeps. Do we need it once we start utilize CentralPackageTransitivePinningEnabled?
> 
> Historically, it prevented me at least once to merge wrong PR. If you see possibility to still have it (this is pretty new feature in .NET) it will be great. I think, it can be done in separate PR, if you can commit to finish it;

I don't care if it is a part of this PR or not. If we decide to handle it in a separate PR, I want an issue submitted to track it.

iskiselev commented:
> > Decide if we need any replacement for deleted TransientDependenciesTests.DefinedTransientDeps_Are_MatchingGeneratedDeps. Do we need it once we start utilize CentralPackageTransitivePinningEnabled?
> 
> Historically, it prevented me at least once to merge wrong PR. If you see possibility to still have it (this is pretty new feature in .NET) it will be great. I think, it can be done in separate PR, if you can commit to finish it;

@Kielek and @nrcventura. I've added new test `ValidateTransientDependenciesPackageVersionsDefined` and think that together with `GeneratorDiscoversTransientDependencies` we have not loss any coverage with `DefinedTransientDeps_Are_MatchingGeneratedDeps` removal.

Previously, every transisient dependency have to be declared twice: as `PackageReference` and `PackageVersion`. `PackageVersion` alone was not enough to change version of transient dependency. With `CentralPackageTransitivePinningEnabled` we don't need to add `PackageReference`, as `PackageVersion` alone enough even for transient dependency.

Removed test logic: 
(1) read list of transient dependency (`PackageReference`) from project.
(2) remove transient dependency (`PackageReference`) from project.
(3) build list of transient dependencies
(4) validate that list from (1) and (3) are the same.

So, test validated:
(1) that dependency list generator is working and produce some set of dependencies
(2) we have not forget to run build and all dependencies in fact listed as `PackageReference` in project

Now:
- GeneratorDiscoversTransientDependencies validates that `System.Memory` transient dependency was discovered - so our dependency generator works and produce some set of dependencies (1). 
- ValidateTransientDependenciesPackageVersionsDefined validates that for every dependency dicovered by generator we have defined `PackageReference` in `Directory.Packages.props` hierarchy (2).

So, both things that were validated before are still validated now.
